File: app/src/main/java/pl/gocards/ui/cards/list/learning_progress/LearningProgressListCardsAdapter.java
package pl.gocards.ui.cards.list.learning_progress;

import android.view.View;

import androidx.annotation.AttrRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;

import com.google.android.material.color.MaterialColors;

import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import pl.gocards.R;
import pl.gocards.db.storage.DatabaseException;
import pl.gocards.room.entity.app.AppConfig;
import pl.gocards.ui.cards.list.search.SearchListCardsAdapter;
import pl.gocards.ui.cards.list.standard.CardViewHolder;

/**
 * C_R_05 Show card status: disabled, forgotten on the right/left edge bar.
 * @author Grzegorz Ziemski
 */
public class LearningProgressListCardsAdapter extends SearchListCardsAdapter {

    private Set<Integer> disabledCards;

    private Set<Integer> forgottenCards;

    private Set<Integer> rememberedCards;

    private String leftEdgeBar;

    private String rightEdgeBar;

    public LearningProgressListCardsAdapter(
            @NonNull LearningProgressListCardsActivity activity
    ) throws DatabaseException {
        super(activity);
    }

    /* -----------------------------------------------------------------------------------------
     * OnCreate
     * ----------------------------------------------------------------------------------------- */

    protected @NonNull Completable beforeOnStart() {
        return Completable.mergeArray(
                super.beforeOnStart(),
                loadLeftEdgeBarConfig(),
                loadRightEdgeBarConfig()
        );
    }

    @NonNull
    private Completable loadLeftEdgeBarConfig() {
        return getAppDb().appConfigRxDao().load(
                AppConfig.LEFT_EDGE_BAR,
                AppConfig.LEFT_EDGE_BAR_DEFAULT,
                // Do not use runOnUiThread, otherwise null is read by other Rx.
                s -> leftEdgeBar = s
        ).ignoreElements();
    }

    @NonNull
    private Completable loadRightEdgeBarConfig() {
        return getAppDb().appConfigRxDao().load(
                AppConfig.RIGHT_EDGE_BAR,
                AppConfig.RIGHT_EDGE_BAR_DEFAULT,
                // Do not use runOnUiThread, otherwise null is read by other Rx.
                s -> rightEdgeBar = s
        ).ignoreElements();
    }

    @Override
    public void onBindViewHolder(@NonNull CardViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);
        setCardStatusEdgeBar((LearningProgressViewHolder) holder, position);
    }

    @UiThread
    private void setCardStatusEdgeBar(@NonNull LearningProgressViewHolder holder, int position) {
        if (AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS.equals(leftEdgeBar)) {
            View view = holder.getLeftEdgeBarView();
            setCardStatusEdgeBar(view, position);
        }
        if (AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS.equals(rightEdgeBar)) {
            View view = holder.getRightEdgeBarView();
            setCardStatusEdgeBar(view, position);
        }
    }

    @UiThread
    @SuppressWarnings("UnnecessaryReturnStatement")
    private void setCardStatusEdgeBar(@NonNull View view, int position) {
        int cardId = Objects.requireNonNull(getItem(position).getId());
        view.setBackgroundColor(0);
        if (setBackgroundColorIfContainsCard(view, disabledCards, cardId, R.attr.colorItemDisabledCard)) return;
        if (setBackgroundColorIfContainsCard(view, forgottenCards, cardId, R.attr.colorItemForgottenCard)) return;
        if (setBackgroundColorIfContainsCard(view, rememberedCards, cardId, R.attr.colorItemRememberedCards)) return;
    }

    @UiThread
    protected boolean setBackgroundColorIfContainsCard(
            @NonNull View itemView,
            @Nullable Set<Integer> cards,
            int cardId,
            @AttrRes int colorAttributeResId
    ) {
        if (cards != null && cards.contains(cardId)) {
            itemView.setBackgroundColor(MaterialColors.getColor(itemView, colorAttributeResId));
            return true;
        }
        return false;
    }

    /* -----------------------------------------------------------------------------------------
     * Items actions
     * ----------------------------------------------------------------------------------------- */

    @Override
    protected @NonNull Completable beforeLoadItems() {
        return Completable.fromAction(() -> {
            if (isShownLearningStatus()) {
                Completable.mergeArray(
                        super.beforeLoadItems(),
                        loadCardLearningProgress()
                ).blockingSubscribe();
            } else {
                super.beforeLoadItems().blockingSubscribe();
            }
        });
    }

    @NonNull
    private Completable loadCardLearningProgress() {
        return Observable.merge(
                getDisabledCards().toObservable(),
                getForgottenCards().toObservable(),
                getRememberedCards().toObservable()
        ).ignoreElements();
    }

    private boolean isShownLearningStatus() {
        return AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS.equals(leftEdgeBar)
                || AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS.equals(rightEdgeBar);
    }

    private Maybe<List<Integer>> getDisabledCards() {
        return getDeckDb().cardRxDao()
                .findIdsByDisabledTrueCards()
                .doOnSuccess(cardIds -> runOnUiThread(() -> disabledCards = new HashSet<>(cardIds)));
    }

    private Maybe<List<Integer>> getForgottenCards() {
        return getDeckDb().cardLearningProgressRxDao()
                .findCardIdsByForgotten()
                .doOnSuccess(cardIds -> runOnUiThread(() -> forgottenCards = new HashSet<>(cardIds)));
    }

    private Maybe<List<Integer>> getRememberedCards() {
        return getDeckDb().cardLearningProgressRxDao()
                .findCardIdsByRemembered()
                .doOnSuccess(cardIds -> runOnUiThread(() -> rememberedCards = new HashSet<>(cardIds)));
    }

    /* -----------------------------------------------------------------------------------------
     * Get/sets
     * ----------------------------------------------------------------------------------------- */

    @NonNull
    @Override
    public LearningProgressListCardsActivity requireActivity() {
        return (LearningProgressListCardsActivity) super.requireActivity();
    }

    public String getLeftEdgeBar() {
        return leftEdgeBar;
    }

    public String getRightEdgeBar() {
        return rightEdgeBar;
    }
}

File: app/src/main/java/pl/gocards/ui/cards/list/edge_bar/learning_progress/LearningProgressListCardsAdapter.kt
package pl.gocards.ui.cards.list.edge_bar.learning_progress

import android.view.View
import androidx.annotation.UiThread
import androidx.compose.material3.SnackbarHostState
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.CoroutineScope
import pl.gocards.room.entity.app.AppConfig
import pl.gocards.ui.cards.list.ListCardsActivity
import pl.gocards.ui.cards.list.display.CardViewHolder
import pl.gocards.ui.cards.list.search.SearchListCardsAdapter
import pl.gocards.ui.cards.list.search.SearchListCardsViewModel
import pl.gocards.ui.cards.list.select.SelectCardsViewModel
import pl.gocards.ui.theme.ExtendedColors


/**
 * C_R_05 Show card status: disabled, forgotten on the right/left edge bar.
 * @author Grzegorz Ziemski
 */
open class LearningProgressListCardsAdapter(
    viewModel: SearchListCardsViewModel,
    selectViewModel: SelectCardsViewModel,
    private val learningProgressViewModel: LearningProgressViewModel,
    snackbarHostState: SnackbarHostState,
    colors: ExtendedColors,
    activity: ListCardsActivity,
    owner: LifecycleOwner = activity,
    scope: CoroutineScope = owner.lifecycleScope
): SearchListCardsAdapter(
    viewModel,
    selectViewModel,
    snackbarHostState,
    colors,
    activity,
    scope
) {

    override fun loadCards() {
        learningProgressViewModel.loadCards {
            super.loadCards()
        }
    }

    override fun onBindViewHolder(holder: CardViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        setStatusEdgeBar(holder, position)
    }

    @UiThread
    private fun setStatusEdgeBar(holder: CardViewHolder, position: Int) {
        if (AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS == learningProgressViewModel.leftEdgeBar) {
            val view = holder.getLeftEdgeBarView()
            setStatusEdgeBar(view, position)
        }
        if (AppConfig.EDGE_BAR_SHOW_LEARNING_STATUS == learningProgressViewModel.rightEdgeBar) {
            val view = holder.getRightEdgeBarView()
            setStatusEdgeBar(view, position)
        }
    }

    @UiThread
    private fun setStatusEdgeBar(view: View, position: Int) {
        val cardId = getCard(position).id
        view.setBackgroundColor(0)
        if (setColorIfContains(
                view,
                learningProgressViewModel.disabledCards,
                cardId,
                colors.colorItemDisabledCard
            )
        ) return
        if (setColorIfContains(
                view,
                learningProgressViewModel.forgottenCards,
                cardId,
                colors.colorItemForgottenCard
            )
        ) return
        if (setColorIfContains(
                view,
                learningProgressViewModel.rememberedCards,
                cardId,
                colors.colorItemRememberedCards
            )
        ) return
    }

    @UiThread
    protected fun setColorIfContains(
        itemView: View,
        cards: Set<Int>,
        cardId: Int,
        color: Color
    ): Boolean {
        if (cards.contains(cardId)) {
            itemView.setBackgroundColor(color.toArgb())
            return true
        }
        return false
    }
}